File: Benchmarks/Core/Scene/SceneFactory.cs
using Benchmarks.Core.Collections;
using Benchmarks.Core.Components;
using Benchmarks.Core.Entity;
using Benchmarks.Core.Types;

namespace Benchmarks.Core.Scene
{
    /// <summary>
    /// Factory que crea escenas de test con entidades realistas para benchmarking.
    /// Distribucion: 70% mundo, 10% NPCs animados, 10% UI, 5% shaders, 5% jugadores.
    /// </summary>
    public static class SceneFactory
    {
        public static Scene CreateTestScene(int entityCount)
        {
            EventManager.Reset();
            var scene = new Scene();
            PopulateEntities(scene.EntityManager, entityCount);
            scene.Initialize();
            return scene;
        }

        public static EntityManager CreateTestEntityManager(int entityCount)
        {
            var entityManager = new EntityManager();
            PopulateEntities(entityManager, entityCount);
            return entityManager;
        }

        private static void PopulateEntities(EntityManager em, int entityCount)
        {
            var rng = new Random(42); // Deterministic seed

            int worldEntities = (int)(entityCount * 0.70);
            int npcEntities = (int)(entityCount * 0.10);
            int uiEntities = (int)(entityCount * 0.10);
            int shaderEntities = (int)(entityCount * 0.05);
            int playerEntities = entityCount - worldEntities - npcEntities - uiEntities - shaderEntities;

            int id = 0;

            // 70% - World entities: Transform + Sprite + RigidBody
            for (int i = 0; i < worldEntities; i++)
            {
                var entity = new Yotsuba(id) { Name = $"World_{id}" };
                em.AddEntity(ref entity);

                em.AddTransformComponent(entity, new TransformComponent(
                    new Vector3(rng.Next(-1000, 1000), rng.Next(-1000, 1000), rng.NextSingle()),
                    new Vector3(rng.Next(1, 4) * 32, rng.Next(1, 4) * 32, 0),
                    rng.NextSingle() * 2 + 0.5f,
                    SpriteEffects.None,
                    Color.White
                ));

                em.AddSpriteComponent(entity, new SpriteComponent2D(
                    rng.Next(1, 100),
                    new Rectangle(0, 0, 32, 32)
                ) { IsVisible = true });

                var gameType = rng.Next(2) == 0 ? GameType.TopDown : GameType.Platform;
                em.AddRigidbodyComponent(entity, new RigidBodyComponent2D(gameType, MassLevel.Collision)
                {
                    Velocity = new Vector3(rng.NextSingle() * 2 - 1, rng.NextSingle() * 2 - 1, 0)
                });

                id++;
            }

            // 10% - NPC entities: Transform + Sprite + Animation + RigidBody
            for (int i = 0; i < npcEntities; i++)
            {
                var entity = new Yotsuba(id) { Name = $"NPC_{id}" };
                em.AddEntity(ref entity);

                em.AddTransformComponent(entity, new TransformComponent(
                    new Vector3(rng.Next(-1000, 1000), rng.Next(-1000, 1000), rng.NextSingle()),
                    new Vector3(64, 64, 0),
                    1f, SpriteEffects.None, Color.White
                ));

                em.AddSpriteComponent(entity, new SpriteComponent2D(
                    rng.Next(1, 100),
                    new Rectangle(0, 0, 64, 64)
                ) { IsVisible = true });

                var animComp = new AnimationComponent2D();
                var idleAnim = new AnimationData
                {
                    CurrentFrame = 0,
                    TotalFrames = 8,
                    FrameTime = 0.1f,
                    FrameWidth = 64,
                    FrameHeight = 64,
                    TextureId = rng.Next(1, 100),
                    Loop = true,
                };
                animComp.AddAnimation(AnimationType.Idle, idleAnim);

                var walkAnim = new AnimationData
                {
                    CurrentFrame = 0,
                    TotalFrames = 6,
                    FrameTime = 0.12f,
                    FrameWidth = 64,
                    FrameHeight = 64,
                    TextureId = rng.Next(1, 100),
                    Loop = true,
                };
                animComp.AddAnimation(AnimationType.Walk, walkAnim);
                animComp.ActivateAnimation(AnimationType.Idle);

                em.AddAnimationComponent(entity, animComp);

                em.AddRigidbodyComponent(entity, new RigidBodyComponent2D(GameType.TopDown, MassLevel.Collision)
                {
                    Velocity = new Vector3(rng.NextSingle() - 0.5f, rng.NextSingle() - 0.5f, 0)
                });

                id++;
            }

            // 10% - UI entities: Transform + Sprite + Button2D
            for (int i = 0; i < uiEntities; i++)
            {
                var entity = new Yotsuba(id) { Name = $"UI_{id}" };
                em.AddEntity(ref entity);

                int bx = rng.Next(0, 1800);
                int by = rng.Next(0, 1000);
                int bw = rng.Next(50, 200);
                int bh = rng.Next(30, 80);

                em.AddTransformComponent(entity, new TransformComponent(
                    new Vector3(bx, by, 0.9f),
                    new Vector3(bw, bh, 0),
                    1f, SpriteEffects.None, Color.White
                ));

                em.AddSpriteComponent(entity, new SpriteComponent2D(
                    rng.Next(1, 100),
                    new Rectangle(0, 0, bw, bh)
                ) { IsVisible = true });

                em.AddButtonComponent2D(entity, new ButtonComponent2D
                {
                    IsActive = true,
                    EffectiveArea = new Rectangle(bx, by, bw, bh),
                    Action = () => { _ = 0; }
                });

                id++;
            }

            // 5% - Shader entities: Transform + Sprite + Shader
            for (int i = 0; i < shaderEntities; i++)
            {
                var entity = new Yotsuba(id) { Name = $"Shader_{id}" };
                em.AddEntity(ref entity);

                em.AddTransformComponent(entity, new TransformComponent(
                    new Vector3(rng.Next(-1000, 1000), rng.Next(-1000, 1000), rng.NextSingle()),
                    new Vector3(128, 128, 0),
                    1f, SpriteEffects.None, Color.White
                ));

                em.AddSpriteComponent(entity, new SpriteComponent2D(
                    rng.Next(1, 100),
                    new Rectangle(0, 0, 128, 128)
                ) { IsVisible = true });

                em.AddShaderComponent2D(entity, new ShaderComponent(rng.Next(1, 10), true));

                id++;
            }

            // 5% - Player entities: Transform + Sprite + Input + RigidBody
            for (int i = 0; i < playerEntities; i++)
            {
                var entity = new Yotsuba(id) { Name = $"Player_{id}" };
                em.AddEntity(ref entity);

                em.AddTransformComponent(entity, new TransformComponent(
                    new Vector3(rng.Next(-500, 500), rng.Next(-500, 500), 0.5f),
                    new Vector3(48, 48, 0),
                    1f, SpriteEffects.None, Color.White
                ));

                em.AddSpriteComponent(entity, new SpriteComponent2D(
                    rng.Next(1, 100),
                    new Rectangle(0, 0, 48, 48)
                ) { IsVisible = true });

                var inputComp = new InputComponent()
                {
                    KeyBoard = new Dictionary<ActionEntityInput, int>
                    {
                        { ActionEntityInput.MoveUp, 0 },
                        { ActionEntityInput.MoveDown, 1 },
                        { ActionEntityInput.MoveLeft, 2 },
                        { ActionEntityInput.MoveRight, 3 },
                        { ActionEntityInput.Jump, 4 },
                    }
                };
                inputComp.AddInput(InputInUse.HasKeyboard);
                em.AddInputComponent(entity, inputComp);

                em.AddRigidbodyComponent(entity, new RigidBodyComponent2D(GameType.Platform, MassLevel.Collision)
                {
                    Velocity = new Vector3(0, 0, 0),
                    SPEED = 2.0f,
                    TOP_SPEED = 5.0f,
                });

                // Set first player as camera target
                if (i == 0)
                {
                    em.CameraEntityId = entity.Id;
                    em.CameraZoom = 1.0f;
                }

                id++;
            }
        }
    }
}

File: Benchmarks_FSharp/Core/Scene/SceneFactory.fs
namespace Benchmarks.Core.Scene

open System
open System.Collections.Generic
open Benchmarks.Core.Collections
open Benchmarks.Core.Components
open Benchmarks.Core.Entity
open Benchmarks.Core.Types

module SceneFactory =

    let private populateEntities (em: EntityManager) (entityCount: int) =
        let rng = Random(42)

        let worldEntities = int (float entityCount * 0.70)
        let npcEntities = int (float entityCount * 0.10)
        let uiEntities = int (float entityCount * 0.10)
        let shaderEntities = int (float entityCount * 0.05)
        let playerEntities = entityCount - worldEntities - npcEntities - uiEntities - shaderEntities

        let mutable id = 0

        // 70% - World entities: Transform + Sprite + RigidBody
        for _ = 0 to worldEntities - 1 do
            let mutable entity = Yotsuba(id)
            entity.Name <- $"World_{id}"
            em.AddEntity(&entity)

            em.AddTransformComponent(entity, TransformComponent(
                Vector3(float32 (rng.Next(-1000, 1000)), float32 (rng.Next(-1000, 1000)), rng.NextSingle()),
                Vector3(float32 (rng.Next(1, 4) * 32), float32 (rng.Next(1, 4) * 32), 0f),
                rng.NextSingle() * 2f + 0.5f,
                SpriteEffects.None,
                Color.White))

            em.AddSpriteComponent(entity, SpriteComponent2D(
                rng.Next(1, 100),
                Rectangle(0, 0, 32, 32)))

            let gameType = if rng.Next(2) = 0 then GameType.TopDown else GameType.Platform
            let mutable rb = RigidBodyComponent2D(gameType, MassLevel.Collision)
            rb.Velocity <- Vector3(rng.NextSingle() * 2f - 1f, rng.NextSingle() * 2f - 1f, 0f)
            em.AddRigidbodyComponent(entity, rb)

            id <- id + 1

        // 10% - NPC entities: Transform + Sprite + Animation + RigidBody
        for _ = 0 to npcEntities - 1 do
            let mutable entity = Yotsuba(id)
            entity.Name <- $"NPC_{id}"
            em.AddEntity(&entity)

            em.AddTransformComponent(entity, TransformComponent(
                Vector3(float32 (rng.Next(-1000, 1000)), float32 (rng.Next(-1000, 1000)), rng.NextSingle()),
                Vector3(64f, 64f, 0f),
                1f, SpriteEffects.None, Color.White))

            em.AddSpriteComponent(entity, SpriteComponent2D(
                rng.Next(1, 100),
                Rectangle(0, 0, 64, 64)))

            let mutable animComp = AnimationComponent2D()
            let mutable idleAnim = Unchecked.defaultof<AnimationData>
            idleAnim.CurrentFrame <- 0
            idleAnim.TotalFrames <- 8
            idleAnim.FrameTime <- 0.1f
            idleAnim.FrameWidth <- 64
            idleAnim.FrameHeight <- 64
            idleAnim.TextureId <- rng.Next(1, 100)
            idleAnim.Loop <- true
            animComp.AddAnimation(AnimationType.Idle, idleAnim)

            let mutable walkAnim = Unchecked.defaultof<AnimationData>
            walkAnim.CurrentFrame <- 0
            walkAnim.TotalFrames <- 6
            walkAnim.FrameTime <- 0.12f
            walkAnim.FrameWidth <- 64
            walkAnim.FrameHeight <- 64
            walkAnim.TextureId <- rng.Next(1, 100)
            walkAnim.Loop <- true
            animComp.AddAnimation(AnimationType.Walk, walkAnim)
            animComp.ActivateAnimation(AnimationType.Idle)

            em.AddAnimationComponent(entity, animComp)

            let mutable rb = RigidBodyComponent2D(GameType.TopDown, MassLevel.Collision)
            rb.Velocity <- Vector3(rng.NextSingle() - 0.5f, rng.NextSingle() - 0.5f, 0f)
            em.AddRigidbodyComponent(entity, rb)

            id <- id + 1

        // 10% - UI entities: Transform + Sprite + Button2D
        for _ = 0 to uiEntities - 1 do
            let mutable entity = Yotsuba(id)
            entity.Name <- $"UI_{id}"
            em.AddEntity(&entity)

            let bx = rng.Next(0, 1800)
            let by = rng.Next(0, 1000)
            let bw = rng.Next(50, 200)
            let bh = rng.Next(30, 80)

            em.AddTransformComponent(entity, TransformComponent(
                Vector3(float32 bx, float32 by, 0.9f),
                Vector3(float32 bw, float32 bh, 0f),
                1f, SpriteEffects.None, Color.White))

            em.AddSpriteComponent(entity, SpriteComponent2D(
                rng.Next(1, 100),
                Rectangle(0, 0, bw, bh)))

            let mutable button = Unchecked.defaultof<ButtonComponent2D>
            button.IsActive <- true
            button.EffectiveArea <- Rectangle(bx, by, bw, bh)
            button.Action <- Action(fun () -> ignore 0)
            em.AddButtonComponent2D(entity, button)

            id <- id + 1

        // 5% - Shader entities: Transform + Sprite + Shader
        for _ = 0 to shaderEntities - 1 do
            let mutable entity = Yotsuba(id)
            entity.Name <- $"Shader_{id}"
            em.AddEntity(&entity)

            em.AddTransformComponent(entity, TransformComponent(
                Vector3(float32 (rng.Next(-1000, 1000)), float32 (rng.Next(-1000, 1000)), rng.NextSingle()),
                Vector3(128f, 128f, 0f),
                1f, SpriteEffects.None, Color.White))

            em.AddSpriteComponent(entity, SpriteComponent2D(
                rng.Next(1, 100),
                Rectangle(0, 0, 128, 128)))

            em.AddShaderComponent2D(entity, ShaderComponent(rng.Next(1, 10), true))

            id <- id + 1

        // 5% - Player entities: Transform + Sprite + Input + RigidBody
        for i = 0 to playerEntities - 1 do
            let mutable entity = Yotsuba(id)
            entity.Name <- $"Player_{id}"
            em.AddEntity(&entity)

            em.AddTransformComponent(entity, TransformComponent(
                Vector3(float32 (rng.Next(-500, 500)), float32 (rng.Next(-500, 500)), 0.5f),
                Vector3(48f, 48f, 0f),
                1f, SpriteEffects.None, Color.White))

            em.AddSpriteComponent(entity, SpriteComponent2D(
                rng.Next(1, 100),
                Rectangle(0, 0, 48, 48)))

            let mutable inputComp = Unchecked.defaultof<InputComponent>
            inputComp.KeyBoard <- Dictionary<ActionEntityInput, int>()
            inputComp.KeyBoard.Add(ActionEntityInput.MoveUp, 0)
            inputComp.KeyBoard.Add(ActionEntityInput.MoveDown, 1)
            inputComp.KeyBoard.Add(ActionEntityInput.MoveLeft, 2)
            inputComp.KeyBoard.Add(ActionEntityInput.MoveRight, 3)
            inputComp.KeyBoard.Add(ActionEntityInput.Jump, 4)
            inputComp.AddInput(InputInUse.HasKeyboard)
            em.AddInputComponent(entity, inputComp)

            let mutable rb = RigidBodyComponent2D(GameType.Platform, MassLevel.Collision)
            rb.Velocity <- Vector3(0f, 0f, 0f)
            rb.SPEED <- 2.0f
            rb.TOP_SPEED <- 5.0f
            em.AddRigidbodyComponent(entity, rb)

            if i = 0 then
                em.CameraEntityId <- entity.Id
                em.CameraZoom <- 1.0f

            id <- id + 1

    let CreateTestScene(entityCount: int) =
        EventManager.Reset()
        let scene = Scene()
        populateEntities scene.EntityManager entityCount
        scene.Initialize()
        scene

    let CreateTestEntityManager(entityCount: int) =
        let entityManager = EntityManager()
        populateEntities entityManager entityCount
        entityManager
